[invoke_subgraph] Don't run the graph twice when autograd enabled

angelayi opened the pull request:
In the [previous PR](https://github.com/pytorch/pytorch/pull/167231/files#diff-e2b74af5d8b538a7d07d18507d27010703742ddad5f819992b55f5abc6d9a502R964-R966) we found that the autograd eager impl of invoke_subgraph calls the subgraph twice. If the subgraph contains effects then effects will be run twice, which is bad. This PR fixes the issue by getting the output metadata from `subgraph`'s `node.meta` if it exists.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168389
* #167364
* #167363
* __->__ #167245
* #167231

Differential Revision: [D87392740](https://our.internmc.facebook.com/intern/diff/D87392740)

angelayi committed:
[invoke_subgraph] Don't run the graph twice when autograd enabled

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Don't run the graph twice when autograd enabled"




[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Don't run the graph twice when autograd enabled"

In the [previous PR](https://github.com/pytorch/pytorch/pull/167231/files#diff-e2b74af5d8b538a7d07d18507d27010703742ddad5f819992b55f5abc6d9a502R964-R966) we found that the autograd eager impl of invoke_subgraph calls the subgraph twice. If the subgraph contains effects then effects will be run twice, which is bad. This PR fixes the issue by getting the output metadata from `subgraph`'s `node.meta` if it exists.




[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Don't run the graph twice when autograd enabled"

In the [previous PR](https://github.com/pytorch/pytorch/pull/167231/files#diff-e2b74af5d8b538a7d07d18507d27010703742ddad5f819992b55f5abc6d9a502R964-R966) we found that the autograd eager impl of invoke_subgraph calls the subgraph twice. If the subgraph contains effects then effects will be run twice, which is bad. This PR fixes the issue by getting the output metadata from `subgraph`'s `node.meta` if it exists.




[ghstack-poisoned]

anijain2305 reviewed (approved):
Nice!

angelayi committed:
Update on "[invoke_subgraph] Don't run the graph twice when autograd enabled"

In the [previous PR](https://github.com/pytorch/pytorch/pull/167231/files#diff-e2b74af5d8b538a7d07d18507d27010703742ddad5f819992b55f5abc6d9a502R964-R966) we found that the autograd eager impl of invoke_subgraph calls the subgraph twice. If the subgraph contains effects then effects will be run twice, which is bad. This PR fixes the issue by getting the output metadata from `subgraph`'s `node.meta` if it exists.




[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Don't run the graph twice when autograd enabled"

In the [previous PR](https://github.com/pytorch/pytorch/pull/167231/files#diff-e2b74af5d8b538a7d07d18507d27010703742ddad5f819992b55f5abc6d9a502R964-R966) we found that the autograd eager impl of invoke_subgraph calls the subgraph twice. If the subgraph contains effects then effects will be run twice, which is bad. This PR fixes the issue by getting the output metadata from `subgraph`'s `node.meta` if it exists.




[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Don't run the graph twice when autograd enabled"

In the [previous PR](https://github.com/pytorch/pytorch/pull/167231/files#diff-e2b74af5d8b538a7d07d18507d27010703742ddad5f819992b55f5abc6d9a502R964-R966) we found that the autograd eager impl of invoke_subgraph calls the subgraph twice. If the subgraph contains effects then effects will be run twice, which is bad. This PR fixes the issue by getting the output metadata from `subgraph`'s `node.meta` if it exists.




[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Don't run the graph twice when autograd enabled"

In the [previous PR](https://github.com/pytorch/pytorch/pull/167231/files#diff-e2b74af5d8b538a7d07d18507d27010703742ddad5f819992b55f5abc6d9a502R964-R966) we found that the autograd eager impl of invoke_subgraph calls the subgraph twice. If the subgraph contains effects then effects will be run twice, which is bad. This PR fixes the issue by getting the output metadata from `subgraph`'s `node.meta` if it exists.




[ghstack-poisoned]

angelayi commented:
@pytorchbot merge -i

angelayi commented:
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D87392740).

angelayi commented:
@pytorchbot merge -f "failure looks unrelated"

yangw-dev commented:
@pytorchbot revert -m "the base pr is broken  internal tests in the stack" -c ghfirst

angelayi committed:
Update on "[invoke_subgraph] Don't run the graph twice when autograd enabled"

In the [previous PR](https://github.com/pytorch/pytorch/pull/167231/files#diff-e2b74af5d8b538a7d07d18507d27010703742ddad5f819992b55f5abc6d9a502R964-R966) we found that the autograd eager impl of invoke_subgraph calls the subgraph twice. If the subgraph contains effects then effects will be run twice, which is bad. This PR fixes the issue by getting the output metadata from `subgraph`'s `node.meta` if it exists.


Differential Revision: [D87392740](https://our.internmc.facebook.com/intern/diff/D87392740)

[ghstack-poisoned]
